### PR TITLE
Add atbash-cipher

### LINKED ISSUE #164: Add Practice Exercise: `atbash-cipher`
Towards https://github.com/exercism/cairo/issues/113

Problem specification:
https://github.com/exercism/problem-specifications/tree/main/exercises/atbash-cipher
    
TODO:
    
- generate the exercise with `./bin/fetch-configlet && ./bin/configlet create --practice-exercise atbash-cipher`
- check if any of the existing tracks (examples below) implement the exercise and you can reuse the implementation:
    - Go: https://github.com/exercism/go/tree/main/exercises/practice
    - Rust: https://github.com/exercism/rust/tree/main/exercises/practice
    - Python: https://github.com/exercism/python/tree/main/exercises/practice
- use exercise's _canonical-data.json_ to see which test cases to implement:
    https://github.com/exercism/problem-specifications/blob/main/exercises/atbash-cipher/canonical-data.json

### PR DESCRIPTION
**Pull request by BNAndras**
Closes #164

### CODE REVIEW

File: exercises/practice/atbash-cipher/.docs/instructions.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+```text
+Plain:  abcdefghijklmnopqrstuvwxyz
+Cipher: zyxwvutsrqponmlkjihgfedcba
+```
+
+It is a very weak cipher because it only has one possible key, and it is a simple mono-alphabetic substitution cipher.
+However, this may not have been an issue in the cipher's time.
+
+Ciphertext is written out in groups of fixed length, the traditional group size being 5 letters, leaving numbers unchanged, and punctuation is excluded.
```
Comment (0xNeshi) [inline exercises/practice/atbash-cipher/.docs/instructions.md:18]: This is a very weird definition of atbash cipher; it generally just ignores punctuation and doesn't group letters in any way.
I opened a topic on the official forum about this https://forum.exercism.org/t/atbash-cipher-test-cases-might-be-wrong/12959

File: exercises/practice/atbash-cipher/.meta/example.cairo [hunk 2]
```diff
@@ -0,0 +1,78 @@
+pub fn encode(phrase: ByteArray) -> ByteArray {
+    let mut result = "";
+
+    let mut i = 0;
+    let mut chunk_size = 0;
+    while i < phrase.len() {
+        let chr = phrase[i];
+
+        if is_whitespace(chr) || is_punctuation(chr) {
+            i += 1;
```
Comment (0xNeshi) [inline exercises/practice/atbash-cipher/.meta/example.cairo:10]: nit: you can increment `i` above the if-statement, which would also remove the need to increment it at the bottom of the loop ([line 26](https://github.com/exercism/cairo/pull/274/files#diff-2a9522c77c2a98f38d12114741a42f53d3b85b61e04885c708df2c5263464aadR26)).

File: exercises/practice/atbash-cipher/.meta/example.cairo [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+fn is_punctuation(chr: u8) -> bool {
+    (chr >= '!' && chr <= '/')
+        || (chr >= ':' && chr <= '@')
+        || (chr >= '[' && chr <= '`')
+        || (chr >= '{' && chr <= '~')
+}
+
+fn lowercase(chr: u8) -> u8 {
+    if chr >= 'A' && chr <= 'Z' {
+        chr + 'a' - 'A'
```
Comment (0xNeshi) [inline exercises/practice/atbash-cipher/.meta/example.cairo:74]:
```suggestion
        chr + ('a' - 'A') // add the distance between lowercase and uppercase letters
```
nit: adds a bit more clarity, as the logic on its face looks weird.

File: exercises/practice/atbash-cipher/.meta/example.cairo [hunk 4]
```diff
@@ -0,0 +1,78 @@
+pub fn encode(phrase: ByteArray) -> ByteArray {
+    let mut result = "";
+
+    let mut i = 0;
+    let mut chunk_size = 0;
+    while i < phrase.len() {
+        let chr = phrase[i];
+
+        if is_whitespace(chr) || is_punctuation(chr) {
```
Comment (0xNeshi) [inline exercises/practice/atbash-cipher/.meta/example.cairo:9]:
```suggestion
        if !is_letter(chr) {
```
nit: we can ignore _all_ non-letter chars. Also makes the `is_punctuation` fn redundant.

**Review by 0xNeshi — CHANGES_REQUESTED** [review]
We can merge this despite the results of the test-cases-related topic https://forum.exercism.org/t/atbash-cipher-test-cases-might-be-wrong/12959

If there comes a need to change the tests, will create a separate issue

File: exercises/practice/atbash-cipher/.docs/instructions.md (hunk 1 above)
Comment (BNAndras) [reply exercises/practice/atbash-cipher/.docs/instructions.md:18]: If you're proposing a change to the canonical test suite, I'd be mindful that that problem spec has been around for ten years and has been implemented across 60+ tracks. It'd take some time and work to first sync the changes to the tracks, and then we'll need to rerun all the completed solutions. All solutions would fail if we're no longer grouping strings in five letter chunks. So you'll get some pushback for sure.

File: exercises/practice/atbash-cipher/.meta/example.cairo (hunk 2 above)
Comment (BNAndras) [reply exercises/practice/atbash-cipher/.meta/example.cairo:10]: Fair. I thought about that, but it felt weird to increment my counter before I did anything. It'd make the code more clear though.

File: exercises/practice/atbash-cipher/.meta/example.cairo (hunk 3 above)
Comment (BNAndras) [reply exercises/practice/atbash-cipher/.meta/example.cairo:74]: I'll pull that into a helper function instead.

File: exercises/practice/atbash-cipher/.docs/instructions.md (hunk 1 above)
Comment (BNAndras) [reply exercises/practice/atbash-cipher/.docs/instructions.md:18]: It's not that we can't change the test suite, but we should have a good reason like there's an edge case to be tested that might be significant for multiple tracks.

Comment (0xNeshi) [reply exercises/practice/atbash-cipher/.docs/instructions.md:18]: It was just seemed confusing, because I know what the actual atbash cipher requires. That's why I opened a topic on the forum to discuss it, but this won't affect this PR, which will be reviewed with the current problem description in mind.

**Comment by BNAndras** [comment]
All tests pass locally and in the CI but it's reporting a failure.

**Comment by 0xNeshi** [comment]
That's weird :thinking: Investigating

**Comment by 0xNeshi** [comment]
@BNAndras the issue was in the `exit 1` after the timeout of cairo tests - it ran no matter the result of the test (success/fail). Fixed this by simply changing the order of precedence of commands so that `exit 1` runs only if tests _fail_.
Merge https://github.com/exercism/cairo/pull/280 ASAP and we can merge the PR too

**Review by 0xNeshi — APPROVED** [review]
